ext/openssl/ossl_pkey.c

```diff
@@ -498,6 +675,33 @@ ossl_pkey_s_generate_key(int argc, VALUE *argv, VALUE self)
     return pkey_generate(argc, argv, self, 0);
 }
 
+/*
+ * call-seq:
+ *    OpenSSL::PKey.from_data(algo_name, parameters) -> pkey
+ *
+ * Generates a new key based on given key parameters.
+ * NOTE: Requires OpenSSL 3.0 or later.
+ *
+ * The first parameter is the type of the key to create, given as a String, for example RSA, DSA, EC etc.
+ * Second parameter is the parameters to be used for the key.
+ *
+ * For details algorithms and parameters see https://www.openssl.org/docs/man3.0/man3/EVP_PKEY_fromdata.html
+ *
+ * == Example
+ *   pkey = OpenSSL::PKey.from_data("RSA", n: 3161751493, e: 65537, d: 2064855961)
+ *   pkey.private? #=> true
+ *   pkey.public_key #=> #<OpenSSL::PKey::RSA...
+ */
+static VALUE
+ossl_pkey_s_from_data(int argc, VALUE *argv, VALUE self)
+{
+#if OSSL_OPENSSL_PREREQ(3, 0, 0)
+    return pkey_from_data(argc, argv, self);
+#else
+    rb_raise(ePKeyError, "OpenSSL::PKey.from_data requires OpenSSL 3.0 or later");
+#endif
+}
```

Comment:
We haven't been very consistent on this, but I think it might be a better design to not define this method for unsupported OpenSSL versions.

I guess a user code that must work with both OpenSSL 1.1 and 3.x will have to check the availability of this method, so it would be easier that way.